### ISSUE TITLE
Reuse of mode containing "starts: { endWithParent: true }" not possible.

### ISSUE THREAD

**Issue by nknapp** [highlightjs/highlight.js#2192]
I'm still doing my analysis, but maybe anyone already has an idea, so I write everything down here.

In the new rules for the Handlebars language, I found that reusing some mode-definitions did not work because highlight somehow modifies them in context to its parent mode.

I have created a minimal test-case to reproduce  this issue:

```js
let hljs = require('../../build');

describe("bugs", function () {
  describe("modes containing 'endsWithParent'", function () {
    it("should be allowed to be reused", function () {
      hljs.registerLanguage('test-language', function (hljs) {

        const TAG_CONTENTS = {className: 'name', begin: /\w+/, starts: {endsWithParent: true}};

        const PAREN_TAG = {className: 'tag', begin: /\(/, end: /\)/, contains: [TAG_CONTENTS]};
        const SQUARE_BRACKET_TAG = {className: 'tag', begin: /\[/, end: /\]/, contains: [TAG_CONTENTS]};

        return {
          contains: [PAREN_TAG, SQUARE_BRACKET_TAG]
        };
      });

      hljs.highlight('test-language', '(abc 123) [abc 123] (abc 123)').value
        .should.equal(
        '<span class="hljs-tag">(<span class="hljs-name">abc</span> 123)</span> ' +
        '<span class="hljs-tag">[<span class="hljs-name">abc</span> 123]</span> ' +
        '<span class="hljs-tag">(<span class="hljs-name">abc</span> 123)</span>'
      )
    })
  })
})

```

The expected output ("should.equal" above) is 

```html
<span class="hljs-tag">(<span class="hljs-name">abc</span> 123)</span> <span class="hljs-tag">[<span class="hljs-name">abc</span> 123]</span> <span class="hljs-tag">(<span class="hljs-name">abc</span> 123)</span>
```
but the actual output is 

```html
<span class="hljs-tag">(<span class="hljs-name">abc</span> 123)</span> <span class="hljs-tag">[<span class="hljs-name">abc</span> 123] (abc 123)</span>
```
As you can see, the mode for `[abc 123]` is not terminated after the closing square bracket, but remains until the end of the string.

Looking at the "compiled" language, I can see that the property  `SQUARE_BRACKET_TAG.contains[0].starts.terminator_end` is `/\)/` but it should by `/\]/`

**Comment by yyyc514** [comment]
Can you wrap this up in a PR with just the failing test and I'll take it.  This bugged me from when you first mentioned it. :-)

**Comment by yyyc514** [comment]
Unless you're also wanting to fix it. :-)

**Comment by yyyc514** [comment]
compileMode probably always need to call `inherit` on starts so that the object is copied rather than reused.

**Comment by nknapp** [comment]
I think the problem is that `compileMode` is modifying its parameter. I'll submit a PR with a change that makes my test-case succeed, but a lot of other tests fail. You can take it from there if you like.

It's dinner time soon...

**Comment by yyyc514** [comment]
`parent` is referenced later.

> I'll submit a PR with a change that makes my test-case succeed, but a lot of other tests fail. You can take it from there if you like.

That'd be great.

**Comment by nknapp** [comment]
I am not sure if this really is the problem I had with the handlebars-language, but it might be.

**Comment by yyyc514** [comment]
Well, your problem was similar.  Probably also worth fixing.  I don't think people defining grammars need to think so hard about what is safe to reuse and not.

**Comment by yyyc514** [comment]
Will be closed by https://github.com/highlightjs/highlight.js/pull/2201

**Comment by nknapp** [comment]
Updated PR with modified handlebars definition reproducing the old bug (now fixed) #2207

**Comment by yyyc514** [comment]
Fixed in master!
